pkg/compact/compact.go

```diff
@@ -286,11 +198,127 @@ func (s *Syncer) GarbageCollect(ctx context.Context) error {
+			Name: "thanos_compact_group_compaction_runs_started_total",
+			Help: "Total number of group compaction attempts.",
+		}, []string{"group"}),
+		compactionRunsCompleted: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_compact_group_compaction_runs_completed_total",
+			Help: "Total number of group completed compaction runs. This also includes compactor group runs that resulted with no compaction.",
+		}, []string{"group"}),
+		compactionFailures: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_compact_group_compactions_failures_total",
+			Help: "Total number of failed group compactions.",
+		}, []string{"group"}),
+		verticalCompactions: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_compact_group_vertical_compactions_total",
+			Help: "Total number of group compaction attempts that resulted in a new block based on overlapping blocks.",
+		}, []string{"group"}),
+		garbageCollectedBlocks:  garbageCollectedBlocks,
+		blocksMarkedForDeletion: blocksMarkedForDeletion,
+	}
+}
+
+// Groups returns the compaction groups for all blocks currently known to the syncer.
+// It creates all groups from the scratch on every call.
+func (g *DefaultGrouper) Groups(blocks map[ulid.ULID]*metadata.Meta) (res []*Group, err error) {
+	groups := map[string]*Group{}
+	for _, m := range blocks {
+		groupKey := DefaultGroupKey(m.Thanos)
+		group, ok := groups[groupKey]
+		if !ok {
+			lbls := labels.FromMap(m.Thanos.Labels)
+			group, err = NewGroup(
+				log.With(g.logger, "group", fmt.Sprintf("%d@%v", m.Thanos.Downsample.Resolution, lbls.String()), "groupKey", groupKey),
+				g.bkt,
+				groupKey,
+				lbls,
+				m.Thanos.Downsample.Resolution,
+				g.acceptMalformedIndex,
+				g.enableVerticalCompaction,
+				g.compactions.WithLabelValues(groupKey),
+				g.compactionRunsStarted.WithLabelValues(groupKey),
+				g.compactionRunsCompleted.WithLabelValues(groupKey),
+				g.compactionFailures.WithLabelValues(groupKey),
+				g.verticalCompactions.WithLabelValues(groupKey),
+				g.garbageCollectedBlocks,
+				g.blocksMarkedForDeletion,
+			)
+			if err != nil {
+				return nil, errors.Wrap(err, "create compaction group")
+			}
+			groups[groupKey] = group
+			res = append(res, group)
+		}
+		if err := group.Add(m); err != nil {
+			return nil, errors.Wrap(err, "add compaction group")
+		}
+	}
+	sort.Slice(res, func(i, j int) bool {
+		return res[i].Key() < res[j].Key()
+	})
+	return res, nil
+}
+
 // Group captures a set of blocks that have the same origin labels and downsampling resolution.
 // Those blocks generally contain the same series and can thus efficiently be compacted.
 type Group struct {
 	logger                      log.Logger
 	bkt                         objstore.Bucket
+	key                         string
 	labels                      labels.Labels
 	resolution                  int64
 	mtx                         sync.Mutex
```

Comment:
It's not clear to me why Group needs a mutex in the first place. It is fully constructed in one goroutine, before dispatched to goroutines that do compaction. (Dispatching creates happens-before relation, and group is not modified afterwards).

It's not part of this PR, but I think this mutex can be dropped entirely. (Separate PR is fine as well)